README.txt
============================================================================
jQuery Simple Overlay
Tom McFarlin, http://tommcfarlin.com, @moretom
============================================================================

ABOUT

Simple Overlay is a jQuery plugin for easily including full screen overlays
in your projects. 

It features a variety of configuration options for incorporating multiple
overlays, custom styles, glossy effect, callbacks, and more.

You can apply styles to the element that triggers the overlay using the
'overlay-trigger' class name. The overlay itself given a class name of
'overlay.'

-------------------------------------------

PARAMETERS

color         string
Specifies the hexidecimal or literal value for the color of the overlay.
Defaults to black.

opacity       number
The level of transparency for the overlay. Ranges from 0 to 1.0.
Defaults to 0.5.

effect        string
Dictates how the overlay is displayed and hidden. Can be either none,
fade, or slide.
Defaults to none.

onShow        function
The function called once the overlay is displayed.
Defaults to null.

onHide        function
The function called once the overlay is hidden.
Defaults to null.

closeOnClick  boolean
Dictates whether or not the overlay is closed when it's clicked.
Defaults to false.

glossy        boolean
Specifies whether or not a glossy effect is applied to the overy.
Defaults to false

container     string
The element to which the overlay is applied. Should be an ID.
Defaults to body.

-------------------------------------------

EXAMPLES

Display a semi-transparent red overlay when the first anchor on the
page has been clicked. Display an alert dialog and close the overlay
when it has been clicked.

  $('a:first').overlay(function(evt) {
    $(this).overlay({
      color: 'red',
      closeOnClick: true,
      onShow: function() {
        alert('Hello world!');
      }
    });
  });

Fade a semi-transparent black, glossy overlay into view once the element
having the ID of #trigger has been clicked.

  $('#trigger').click(function() {
    $(this).overlay({
      effect: 'fade',
      glossy: true
    });
  });

Append a new element to the DOM once the default overlay has been
displayed:

  $('#trigger').click(function() {
    $(this).overlay({
      onShow: function() {
        $('body').append(
          $('<div id="new-element">This is my new element!</div>');
        );
      });
    });
  });

-------------------------------------------

CONTACT

Web:      http://tommcfarlin.com
Twitter:  @moretom
Email:    tom@tommcfarlin.com

============================================================================

MIT license:
http://www.opensource.org/licenses/mit-license.php

changelog.txt
= 8/5 =
* [ Issue 1 ] Overlays now correct position themselves in relatively positioned elements.
* Updating the minified build

= 6/27 =
* Fixing a problem with the onHide() callback

= 6/24 = 
* Addressing a problem with Firefox 4 positioning

tests/test.js
/*
 * Simple Overlay Test Suite
 * Unit Tests for Simple Overlay
 * A jQuery Plugin for creating a simple, customizable overlay. Supports multiple instances,
 * custom callbacks, hide on click, glossy effect, and more.
 *
 * Copyright 2011 Tom McFarlin, http://tommcfarlin.com
 * Released under the MIT License
 * More information: http://moreco.de/simple-overlay
*/

/*-------------------------------------------------------*
 * Default Value Tests
 *-------------------------------------------------------*/

module("Default Option Values");
test("color", function() {
	function defaultColorValue() {
		return $.fn.overlay.defaults.color;
	}
	equal(defaultColorValue(), '#000', 'The default color for the overlay.');
});

test("opacity", function() {
	function defaultOpacityValue() {
		return $.fn.overlay.defaults.opacity;
	}
	equal(defaultOpacityValue(), 0.5, 'The level of opacity for the overlay.');
});

test("effect", function() {
	function defaultDisplayEffect() {
		return $.fn.overlay.defaults.effect;
	}
	equal(defaultDisplayEffect(), 'none', 'Verifies that no effect is applied for the default overlay.');
});

test("onShow", function() {
	function defaultOnShowValue() {
		return $.fn.overlay.defaults.onShow;
	}
	equal(defaultOnShowValue(), null, 'Verifies there is no default callback when the overlay is shown.');
});

test("onHide", function() {
	function defaultOnHideValue() {
		return $.fn.overlay.defaults.onHide;
	}
	equal(defaultOnHideValue(), null, 'Verifies there is no default callback when the overlay is hidden.');
});
 
test("closeOnClick", function() {
	function defaultCloseOnClickValue() {
		return $.fn.overlay.defaults.closeOnClick;
	}
	equal(defaultCloseOnClickValue(), false, 'Verifies the modal is not closed by clicking by default.');
});

test("glossy", function() {
	function defaultGlossyValue() {
		return $.fn.overlay.defaults.glossy;
	}
	equal(defaultGlossyValue(), false, 'Verifies there is no gloss effect applied to the default overlay.');
});

test("container", function() {
	function defaultContainer() {
		return $.fn.overlay.defaults.container;
	}
	equal(defaultContainer(), 'body', 'Verifies the body element is the default container.');
});

/*-------------------------------------------------------*
 * Specific Value Tests
 *-------------------------------------------------------*/

module("Specific Option Values");
test("color", function() {
	function colorValue() {
    $('#overlay-trigger').click(function() {
      $(this).overlay({
        color: '#fff',
        closeOnClick: true
      });
    });
    $('#overlay-trigger').trigger('click');
    return $('.overlay:first').css('display', 'none').css('background-color');
	}
	equal(colorValue(), 'rgb(255, 255, 255)', 'The specified color for the overlay.');
});

test("opacity", function() {
	function opacityValue() {
    $('.overlay:first').trigger('click');     // clears the last overlay
		$('#overlay-trigger').click(function() {
      $(this).overlay({
        opacity: 1
      });
    });
    $('#overlay-trigger').trigger('click');
    return $('.overlay:first').css('display', 'none')
      .css('opacity');
  }
	equal(opacityValue(), 1, 'The level of opacity for the overlay.');
});

/*-------------------------------------------------------*
 * Event Tests
 *-------------------------------------------------------*/

asyncTest('onShow', function() {
  
  $('.overlay').trigger('click');
  
  $('#overlay-trigger').click(function() {
    $(this).overlay({
      onShow: function() {
        $('body').addClass('added-by-overlay');
      },
      closeOnClick: true
    });
  });
  
  $('#overlay-trigger').trigger('click');
  
  setTimeout(function() {
    equal($('body').hasClass('added-by-overlay'), true, 'Verifies the onShow callback is properly called');
    start();
  }, 1000);
  
  $('.overlay').trigger('click');
  
});

asyncTest('onHide', function() {

  $('#overlay-trigger').click(function() {
    $(this).overlay({
      onHide: function() {
        $('body').addClass('added-by-overlay');
      },
      closeOnClick: true
    });
  });
  
  $('#overlay-trigger').trigger('click');
  
  setTimeout(function() {
    equal($('body').hasClass('added-by-overlay'), true, 'Verifies the onHide callback is properly called');
    start();
  }, 1000);
  
  $('.overlay').trigger('click');
  
});

asyncTest('closeOnClick', function() {

  $('.overlay').trigger('click').remove();

  $('#overlay-trigger').click(function() {
    $(this).overlay({
      closeOnClick: true
    });
  });

  setTimeout(function() {
  
    $('#overlay-trigger').trigger('click');
    $('.overlay').click();
    
    equal($('.overlay').length === 0, true, 'Verifies the overlay is closed when clicked.');
    start();
  }, 1000);
  
});

asyncTest("glossy", function() {

  $('.overlay').trigger('click');

	$('#overlay-trigger').click(function() {
    $(this).overlay({
      glossy: true,
      closeOnClick: true
    });
  });
  
  setTimeout(function() {
    
    $('#overlay-trigger').click();  
    
    equal($('.overlay').children().length > 0, true, 'Verifies the glossy element is added to the overlay.');
    
    start();
    
    $('.overlay').click();
  
  }, 1000);
  
});

asyncTest("container", function() {

  $('.overlay').trigger('click');

	$('#overlay-trigger').click(function() {
    $(this).overlay({
      glossy: true,
      closeOnClick: true,
      container: '#overlay-container'
    });
  });
  
  setTimeout(function() {
    
    $('#overlay-trigger').click();  
    
    equal($('#overlay-container').children().length > 0, true, 'Verifies the overlay is added to another element rather than the body.');
    
    start();
    
    $('.overlay').click();
  
  }, 1000);
  
});


js/site.js
$(function() {
  
  SyntaxHighlighter.all();
  
  $('#paragraph-trigger').click(function() {
    $(this).overlay({
      closeOnClick: true
    });
  });
  
  var $modal = $('<div>This is my modal.</div>')
    .attr('id', 'modal')
    .css({
      background: '#fff',
    zIndex: 3000,
    padding: '10px',
    width: '640px',
    height: '240px',
    margin: '0 auto',
    opacity: 1,
    position: 'absolute',
    top: '10%',
    left: '10%'
  });
  
  $('#modal-trigger').click(function(evt) {
    evt.preventDefault();
    $(this).overlay({
      effect: 'fade',
      opacity: 0.8,
      closeOnClick: true,
      onShow: function() {
        $('body').append($modal);
      },
      onHide: function() {
        $modal.remove();
      },
    })
  });
  
  $('#anchor-trigger').click(function(evt) {
    evt.preventDefault();
    $(this).overlay({
      color: 'red',
      closeOnClick: true,
      glossy: true
    });
  });  

  $('#photo-trigger').click(function() {
    $(this).overlay({
      color: '#ccc',
      effect: 'fade',
      glossy: true,
      container: '#photo-trigger',
      onShow: function() {
        $(this).click(function(evt) {
          evt.preventDefault();
        }).bind('contextmenu', function(evt) {
          evt.preventDefault();
        });
      }
    });
  });
});

js/jquery.overlay.js
/*
 * jQuery Simple Overlay
 * A jQuery Plugin for creating a simple, customizable overlay. Supports multiple instances,
 * custom callbacks, hide on click, glossy effect, and more.
 *
 * Copyright 2011 Tom McFarlin, http://tommcfarlin.com, @moretom
 * Released under the MIT License
 *
 * http://moreco.de/simple-overlay
 */

(function($) {

	$.fn.overlay = function(options) {

    var opts = $.extend({}, $.fn.overlay.defaults, options);
		return this.each(function(evt) {
      if(!$(this).hasClass('overlay-trigger')) {
        show(create($(this), opts), opts);
      }
		});
    
	}; // end overlay

  /*--------------------------------------------------*
   * helper functions
   *--------------------------------------------------*/
  
  /**
   * Creates the overlay element, applies the styles as specified in the 
   * options, and sets up the event handlers for closing the overlay.
   *
   * opts The plugin's array options.
   */
  function create($src, opts) {
  
    // prevents adding multiple overlays to a container
    $src.addClass('overlay-trigger');
  
    // create the overlay and add it to the dom
    var iTop = 0;
    if($.browser.mozilla && opts.container.toString() === 'body') { 
      iTop = $('html').scrollTop();
    } else {
      iTop = $(opts.container).scrollTop();
    } // end if/else
    
    var overlay = null;
    if($src.css('position') === 'relative') {
    
    	overlay = $('<div></div>')
    	.addClass('overlay')
	      .css({
	        background: opts.color,
	        opacity: opts.opacity,
	        top: $src.offset().top,
	        left: $src.offset().left,
	        width: $src.width(),
	        height: $src.height(),
	        position: 'absolute',
	        zIndex: 1000,
	        display: 'none',
	        overflow: 'hidden'
	      });
    
    } else {
    
	    overlay = $('<div></div>')
	      .addClass('overlay')
	      .css({
	        background: opts.color,
	        opacity: opts.opacity,
	        top: opts.container.toString() === 'body' ? iTop : $(opts.container).offset().top,
	        left: $(opts.container).offset().left,
	        width: opts.container === 'body' ? '100%' : $(opts.container).width(),
	        height: opts.container === 'body' ? '100%' : $(opts.container).height(),
	        position: 'absolute',
	        zIndex: 1000,
	        display: 'none',
	        overflow: 'hidden'
	      });
    
    } // end if/else
    

    // if specified, apply the gloss
    if(opts.glossy) {
      applyGloss(opts, overlay);     
    } // end if
    
    // setup the event handlers for closing the overlay
    if(opts.closeOnClick) {
      $(overlay).click(function() {
        close(overlay, opts);
        $src.removeClass('overlay-trigger');
      });
    } // end if
    
    // finally add the overlay
    $(opts.container).append(overlay);
   
    return overlay;
    
  } // end createOverlay
  
  /**
   * Displays the overlay using the effect specified in the options. Optionally
   * triggers the onShow callback function.
   *
   * opts The plugin's array options.
   */
  function show(overlay, opts) {
    
    switch(opts.effect.toString().toLowerCase()) {
    
      case 'fade':
        $(overlay).fadeIn('fast', opts.onShow);
        break;
      
      case 'slide':
        $(overlay).slideDown('fast', opts.onShow);
        break;
        
      default:
        $(overlay).show(opts.onShow);
        break;
    
    } // end switch/case
    
    $(opts.container).css('overflow', 'hidden');
    
  } // end show
  
  /**
   * Hides the overlay using the effect specified in the options. Optionally
   * triggers the onHide callback function.
   *
   * opts The plugin's array options.
   */
  function close(overlay, opts) {
    
    switch(opts.effect.toString().toLowerCase()) {
        
      case 'fade':
        $(overlay).fadeOut('fast', function() {
          if(opts.onHide) {
            opts.onHide();
          }
          $(this).remove();
        });
        break;
            
      case 'slide':
        $(overlay).slideUp('fast', function() {
          if(opts.onHide) {
            opts.onHide();
          }
          $(this).remove();
        });
        break;
            
      default:
        $(overlay).hide();
        if(opts.onHide) {
          opts.onHide();
        }
        $(overlay).remove();
        break;
            
    } // end switch/case
    
    $(opts.container).css('overflow', 'auto');
    
  } // end close
  
  /**
   * Adds the gloss effect to the overlay.
   *
   * opts     The plugin's options array
   * overlay  The overlay on which the gloss will be applied
   */
  function applyGloss(opts, overlay) {
  
    var gloss = $('<div></div>');
    $(gloss).css({
      background: '#fff',
      opacity: 0.2,
      width: '200%',
      height: '100%',
      position: 'absolute',
      zIndex: 1001,
      msTransform: 'rotate(45deg)',
      webkitTransform: 'rotate(45deg)',
      oTransform: 'rotate(45deg)'
    });
      
    // at the time of development, mozTransform didn't work with >= jQuery 1.5
    if($.browser.mozilla) {
     $(gloss).css('-moz-transform', 'rotate(45deg');
    } // end if
     
    $(overlay).append(gloss);
    
  } // end applyGloss
 
  /*--------------------------------------------------*
   * default settings
   *--------------------------------------------------*/
   
	$.fn.overlay.defaults = {
    color: '#000',
    opacity: 0.5,
    effect: 'none',
    onShow: null,
    onHide: null,
    closeOnClick: false,
    glossy: false,
    container: 'body'
	}; // end defaults

})(jQuery);